SearchKit - Add search templates feature

colemanw opened the pull request:
Overview
----------------------------------------
Search templates are identical to other saved searches, but they cannot be used
as a smart group or placed in an Afform.

They are intended to be cloned and used as the starting place for building a new search.

This adds in the UI a new tab for templates, and a button allowing any existing search
to be saved as a template.

![image](https://github.com/user-attachments/assets/e2abd975-b85e-4e03-bba5-61c47c1a4e57)



Technical Details
------

In the database, the new column is_template is added to the civicrm_saved_search table.

colemanw committed:
SearchKit - Add search templates feature

Search templates are identical to other saved searches, but they cannot be used
as a smart group or placed in an Afform.

They are intended to be cloned and used as the starting place for building a new search.

This adds in the UI a new tab for templates, and a button allowing any existing search
to be saved as a template.

In the database, the new column is_template is added to the civicrm_saved_search table.

colemanw commented:
I've updated [SearchKit Report Starter Pack](https://civicrm.org/extensions/searchkit-report-starter-pack) extension to use the new `is_template` flag, so the easiest way to test this PR is to install that extension along with this patch.

ufundo commented:
Concept makes sense and code looks good.

But on the test site some of the functional restrictions on templates don't seem to work?

I created this template from a search http://core-31528-3wzhm.test-1.civicrm.org:8003/civicrm/admin/search#/edit/13

And I seemed to be able to add a smart group, and to create an afform for that search?

From the code I expected those things to not be allowed for templates, or have I misunderstood?

ufundo commented:
Not blocking, but I think it could be good to have some clear indication when you are editing a template as opposed to a regular search kit. 

Especially when you first click "Save as template" and get taken to an edit screen, it's not very obvious what's happened:
![image](https://github.com/user-attachments/assets/15077e0f-b74c-40ec-8057-81b4e6f81f52)

- could that "(copy)" say "(template)" when it's a template
- maybe the Save button could say "Save template" ?

ufundo commented:
Oh wait - actually. It only shows that "Search Template:" prefix when you first click save as template.

If you then go to Search Templates tab, find a search template, and hit Edit - it goes into normal edit mode, even though you are editing a template. I think that is what allows me to create the smart group etc as well

colemanw commented:
Thanks @ufundo, I had just forgotten to add `'is_template'` to the select clause when loading the savedSearch for edit mode. Fixed now.

ufundo commented:
Ah, definitely better!

> they cannot be used as a smart group or **placed in an Afform**

I still seem to be able to put the template in an afform if I start from the SearchKit edit page:

![image](https://github.com/user-attachments/assets/9dca8f3e-6880-4a6a-8ca9-2840209f08be)
http://core-31528-45rb4.test-1.civicrm.org:8003/civicrm/admin/search#/edit/13

I think that's a blocker because if people can get into this unexpected state through the UI, they inevitably will, and can see that causing weird bugs down the line.

colemanw commented:
Another good catch @ufundo. Fixed, it was just a matter of conditionally showing that dropdown.

ufundo commented:
retest this please

ufundo commented:
Working nicely now.

Final nit:


I think on a template the action wordings are a bit confusing.
![image](https://github.com/user-attachments/assets/b4c7222d-f374-4b49-a636-99ef811aca2b)

For a template it looks like "Clone" creates a non-template - I think that would be good as "Create from template".
"Save as template" meanwhile would make sense to be labelled "Clone" in this context.

I don't think that is a blocker, so I've added merge ready, could be a follow up

colemanw committed:
SearchKit - Fix fetching search template by name

Smarter condition to check whether the record is being fetched by unique identifier
before applying default where clause.

colemanw commented:
Thanks @ufundo I've updated the labels, fixed a bug & added a unit test.

ufundo commented on Civi/Api4/Generic/AbstractGetAction.php:
```diff
@@ -34,13 +34,13 @@ public function selectRowCount() {
    * @throws \CRM_Core_Exception
    */
   public function setDefaultWhereClause() {
-    if (!$this->_itemsToGet('id')) {
```
Is this the best fix - feels like it could affect a *lot* of other API queries?

colemanw replied:
@ufundo I would say yes, under the circumstances it is the best fix. Fetching a record by `name` is basically the same as fetching it by `id` so it should behave the same. I can't see this negatively affecting other api calls, but it should prevent some future headaches.

I say 'under the circumstances' because I think the *best* fix would be to go back in time 5 years and not put in these hard-coded default where clauses in the first place. But short of that...

ufundo replied:
Sorry I don't have the context. I'm also confused at what this defaultWhere clause is. Automatically exclude any records that don't have default value for any fields that have a default value? :thinking: 

Also just to sense check, `_itemsToGet('name')` **doesn't** get populated for queries a where clause like `['name', 'CONTAINS', 'a'] `?

colemanw replied:
Whoops, I spoke too soon. Test failure says it's more complicated. Hang on I have an idea to do something smarter here & check unique indices instead of fieldname guesswork.

colemanw commented:
Merging per review. Let's get this into the RC.
